synapse/dendrite/parser.py

```diff
@@ -0,0 +1,130 @@
+        self._assertOpts()
+        self.svcbus = s_telepath.openurl(self.opts.get('svcbus'))
+        self.proxy = s_service.SvcProxy(self.svcbus)
+        self.coordinator = self.proxy.getNameProxy(self.opts.get('coordinator'))
+        self.jobs = self.proxy.getNameProxy(self.opts.get('jobs'))
+
+    def register(self, mimeType, queue):
+        '''
+        Register a MIME type to queue name with the Dendrite Coordinator.
+
+        As a part of the registration process, a listener is attached to the ``'dendrite:jobs:%s' % queue`` in order
+        to detect when jobs become available for processing.
+
+        Args:
+            mimeType (str):
+                The MIME type as is returned by filemagic with the MAGIC_MIME_TYPE flag set. For more
+                information, see: https://filemagic.readthedocs.io/en/latest/api.html.
+
+            queue (str):
+                The name of the queue for the Coordinator to put jobs on.
+        '''
+        logger.info('registering %s => %s, and svcbus event: %s', mimeType, queue, self._eventName(queue))
+        self.coordinator.register(mimeType, queue)
+        self.svcbus.on(self._eventName(queue), self._processEvent)
+
+    def unregister(self, mimeType, queue):
+        '''
+        Unregister a MIME type to queue name mapping with the Dendrite Coordinator.
+
+        Removes the MIME type to queue name mapping that the Coordinator uses to enqueue jobs. This has a system wide
+        impact and individual Parsers should use care when calling this method, as there may be other parser instances
+        running that expect the registration to exist.
+
+        Args:
+            mimeType (str):
+                The MIME type as is returned by filemagic with the MAGIC_MIME_TYPE flag set. For more
+                information, see: https://filemagic.readthedocs.io/en/latest/api.html.
+
+            queue (str):
+                The name of the queue to unregister.
+        '''
+        logger.info('unregistering %s => %s, and svcbus event: %s', mimeType, queue, self._eventName(queue))
+        self.coordinator.unregister(mimeType, queue)
+        self.svcbus.off(self._eventName(queue), self._processEvent)
+
+    def process(self, queue, job):
+        '''
+        Process a parse job.
+
+        This method invokes the internal method '_process', which should be overridden by sub-classes. The typical
+        implementation should involve actions such as reading the bytes from the Axon service, parse the bytes, and
+        store results in a Cortex. If an exception occurs during the processing and the exception is allowed to
+        propagate out of the process method, the job will be moved to a 'failed' status.
+
+        Args:
+            queue (str):
+                Name of the queue that the job came from.
+
+            job (dict):
+                Dict containing properties of the job.
+        '''
+        self._process(queue, job)
+
+    def processAll(self, queue):
```

Comment:
In order to go to large scale, we will need each host ( actual server ) running a processor (for a given queue) to only be connected to the svcbus/et.al once.  On large systems (with many cores etc) we will likely be running many parsers simultaneously ( the new cluster is 700 cores but only 16 servers ) so maintaining a separate socket for each worker will be a scale problem very quickly.

Additionally, parsers being implemented by 3rd parties means stability problems if they're not being run separately from the python interpreter communicating upstream.  Aka, the _process() method will always need to be run in a sub process of some type.  We've regularly seen 3rd party parser dependencies (in C, which are designed for "good" input) crashing the whole interpreter.  Additionally, without separate processes, we have no way to enforce any kind of timeout on jobs which have gone to lunch ( again, 3rd party parsers, with infinite loop parser bugs etc.. ).

Also, from above, it def seems like we're going to fire a crushing load of a zillion "please go check for a job" requests for every single job we queue....